## Lab Assignment 10: <br>
### Student Name: Ramandeep Kaur
### Student ID: 8976849

This assignment targets to assure that the students understood basic IR concepts.<br>

**Part 1: Implementing an IR System (10/16)**<br>

Consider a collection of 1000 documents and a set of 10 queries. Implement an IR system based on the Vector Space Model (VSM) using TF-IDF weighting.<br>

Dataset:<br>

- 1000 documents (text content for each document)
- 10 sample queries <br>
IR System:<br>

Implement TF-IDF calculation for document-term matrix construction. <br>
Develop a cosine similarity-based retrieval system to rank documents for each query. <br>
Rank the top 10 documents for each query using the IR system. <br>
Evaluation: <br>

Compute Precision at k (P@k) for k=5, k=6, and k=10 for each query.<br>
Calculate Mean Average Precision (MAP) across all queries. <br>
Calculate the Mean Reciprocal Rank (MRR) across all queries.(3) <br><br>
**Part 2: Assessing Inter-Annotator Agreement (6/16)**<br>

Given the relevance assessments by three different annotators for a set of documents:<br>

Annotator 1,2 and 3 relevance assessments<br>

You are expected to: <br>

Compute pairwise Cohen's Kappa values for the annotators' relevance assessments.<br>
Discuss the observed agreement levels among annotators.<br>
Explain how to improve the kappa value if we are not satisfied with the kappa.<br>
Hint: Check Kappa Measure Last Slide Week 11


----------------------------------------------------------------------------------------
- Part 1

In [5]:
documents = [
    "Machine learning is transforming various industries.",
    "Natural language processing helps in text analysis.",
    "AI algorithms can improve decision-making processes.",
    "Data science involves extracting insights from data.",
    "Robotics is a field combining hardware and software.",
    "Deep learning models require large amounts of data.",
    "Blockchain technology secures digital transactions.",
    "Cloud computing offers scalable computing power.",
    "Virtual reality provides immersive experiences.",
    "Augmented reality enhances real-world environments.",
    "Internet of Things connects devices for data exchange.",
    "Biometric authentication ensures secure access.",
    "Quantum computing promises faster computations.",
    "Cybersecurity protects against digital threats.",
    "Ethical considerations in AI development are crucial.",
    "Healthcare benefits from AI-driven diagnostics.",
    "E-commerce relies on personalized recommendation systems.",
    "Autonomous vehicles revolutionize transportation.",
    "Smart cities utilize IoT for efficient operations.",
    "Social media analysis aids in understanding trends.",
    "Predictive analytics anticipates future outcomes.",
    "Remote work trends increase reliance on digital tools.",
    "Energy efficiency through smart grids is vital.",
    "Fintech innovations reshape financial services.",
    "Data privacy concerns arise in the era of big data.",
    "Artificial general intelligence remains a challenge.",
    "Human-computer interaction shapes user experiences.",
    "Big data analytics drives informed decision-making.",
    "Climate change predictions benefit from AI models.",
    "Bioinformatics applies computational techniques to biology.",
    "Robotic process automation streamlines workflows.",
    "Industry 4.0 integrates AI with manufacturing.",
    "Sentiment analysis detects emotions in text data.",
    "Reinforcement learning powers autonomous agents.",
    "Data visualization simplifies complex information.",
    "Edge computing reduces latency in data processing.",
    "Smart farming optimizes agricultural practices.",
    "AI in education enhances personalized learning.",
    "Natural disaster predictions leverage machine learning.",
    "Media recommendation systems personalize content.",
    "Speech recognition enables hands-free interactions.",
    "Genetic algorithms mimic natural selection.",
    "Neural networks simulate the human brain.",
    "Behavioral analytics uncovers patterns in behavior.",
    "Spatial analysis benefits from geospatial data.",
    "Explainable AI improves transparency in models.",
    "AI ethics guide responsible technology development.",
    "Prescriptive analytics offers actionable insights.",
    "Supply chain optimization employs predictive models.",
    "Emotion AI detects emotions in facial expressions.",
    "Distributed ledger technology ensures secure transactions.",
    "Biological data analysis aids in medical research.",
    "Smart grid technology enhances energy distribution.",
    "AI-powered chatbots automate customer support.",
    "Machine translation breaks language barriers.",
    "Personalized medicine tailors treatments to individuals.",
    "Data-driven marketing optimizes customer engagement.",
    "Smart home devices enhance living experiences.",
    "Time series forecasting predicts future trends.",
    "AI-driven creativity challenges human capabilities.",
    "Intelligent document processing automates data extraction.",
    "Graph analytics explores relationships in networks.",
    "Adversarial attacks pose challenges to AI security.",
    "Predictive maintenance minimizes equipment downtime.",
    "Spatial reasoning enhances AI navigation systems.",
    "Privacy-preserving techniques protect sensitive data.",
    "Automated decision-making raises ethical concerns.",
    "Behavioral biometrics verifies user identities.",
    "Explainable recommendations increase user trust.",
    "Quantum machine learning explores quantum states.",
    "Internet censorship detection uses AI techniques.",
    "Smart wearables monitor health and fitness.",
    "Intelligent tutoring systems adapt to student needs.",
    "Graph neural networks model complex relationships.",
    "AI in sports analytics improves performance analysis.",
    "Facial recognition technology raises privacy debates.",
    "Digital twin technology replicates physical systems.",
    "Emotion detection in video content aids analysis.",
    "Neuromorphic computing mimics brain structure.",
    "Robotic surgery enhances precision in operations.",
    "AI-powered language translation assists global communication.",
    "Automated content moderation filters online content.",
    "Network anomaly detection identifies security threats.",
    "Predictive policing uses data to prevent crime.",
    "Quantum cryptography ensures secure communications.",
    "AI-generated art challenges traditional creativity.",
    "Ethical considerations in autonomous vehicles are debated.",
    "Adaptive learning systems personalize educational content.",
    "Biomedical image analysis aids in disease diagnosis.",
    "Explainable computer vision interprets image features.",
    "AI-driven personalization enhances user experiences.",
    "Conversational AI improves human-like interactions.",
    "Federated learning protects individual data privacy.",
    "Human-centered AI design prioritizes user needs.",
    "Quantum annealing solves optimization problems.",
    "Behavior-based authentication enhances security.",
    "AI in music composition transforms creative processes.",
    "Semantic search improves information retrieval.",
    "Recommender systems optimize user preferences.",
    "Information Retrieval systems are outdated without neural networks"
]
print(len(documents))

100


In [6]:
queries = [
    "What are the impacts of AI on healthcare?",
    "How does machine learning improve financial services?",
    "What is the role of AI in autonomous vehicles?",
    "Explain the applications of natural language processing.",
    "How does robotics benefit from AI integration?",
    "What are the challenges in implementing AI in education?",
    "How does data science contribute to climate change predictions?",
    "What are the ethical considerations in AI development?",
    "Explain the significance of AI in smart cities.",
    "How does AI enhance cybersecurity measures?"
]
print(len(queries))

10


In [8]:
ground_truth =[[3, 74, 38, 92, 71, 72, 98, 48, 18, 100],
                [12, 26, 52, 54, 21, 63, 64, 45, 14, 83],
                [58, 77, 62, 31, 6, 61, 36, 96, 85, 18],
                [99, 62, 98, 34, 63, 79, 43, 31, 3, 16],
                [8, 88, 73, 82, 37, 25, 34, 87, 66, 58],
                [90, 40, 84, 64, 34, 2, 73, 23, 59, 89],
                [80, 10, 86, 21, 68, 37, 83, 57, 6, 98],
                [22, 3, 2, 44, 56, 80, 50, 63, 87, 13],
                [1, 18, 22, 44, 99, 72, 24, 95, 10, 87],
                [31, 10, 56, 50, 75, 4, 18, 85, 84, 74]
                           ]
print(len(ground_truth))


10


In [9]:
## TODO: Implement your solution for part 1

import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

# Preprocessing and TF-IDF calculation
tfidf_vectorizer = TfidfVectorizer()
tfidf_matrix = tfidf_vectorizer.fit_transform(documents)
queries_tfidf = tfidf_vectorizer.transform(queries)

# Cosine similarity calculation and ranking
ranked_docs = []
for query_tfidf in queries_tfidf:
    similarities = cosine_similarity(query_tfidf, tfidf_matrix).flatten()
    ranked_indices = np.argsort(similarities)[::-1]  # Sort indices by descending order of similarity
    ranked_docs.append(ranked_indices)

# Evaluation
k_values = [5, 6, 10]
precisions_at_k = []
mean_average_precision = 0
mean_reciprocal_rank = 0

for i, ranked_indices in enumerate(ranked_docs):
    precisions = []
    reciprocal_rank = 0
    relevant_count = 0
    average_precision = 0

    for k in k_values:
        relevant_docs = set(ground_truth[i][:k])
        retrieved_docs = set(ranked_indices[:k])
        precision = len(relevant_docs.intersection(retrieved_docs)) / k
        precisions.append(precision)

        # Calculate Average Precision
        relevant_count += len(relevant_docs)
        retrieved_count = 0
        for j, doc_index in enumerate(ranked_indices[:k]):
            if doc_index in relevant_docs:
                retrieved_count += 1
                average_precision += retrieved_count / (j + 1)

        # Calculate Reciprocal Rank
        if reciprocal_rank == 0:
            for j, doc_index in enumerate(ranked_indices[:k]):
                if doc_index in relevant_docs:
                    reciprocal_rank = 1 / (j + 1)
                    break

    mean_average_precision += average_precision / relevant_count
    mean_reciprocal_rank += reciprocal_rank
    precisions_at_k.append(precisions)

mean_average_precision /= len(queries)
mean_reciprocal_rank /= len(queries)

# Display results
for i, query in enumerate(queries):
    print(f"Query: {query}")
    for j, k in enumerate(k_values):
        print(f"P@{k}: {precisions_at_k[i][j]}")
    print()

print(f"Mean Average Precision (MAP): {mean_average_precision}")
print(f"Mean Reciprocal Rank (MRR): {mean_reciprocal_rank}")


Query: What are the impacts of AI on healthcare?
P@5: 0.0
P@6: 0.0
P@10: 0.0

Query: How does machine learning improve financial services?
P@5: 0.0
P@6: 0.16666666666666666
P@10: 0.1

Query: What is the role of AI in autonomous vehicles?
P@5: 0.0
P@6: 0.0
P@10: 0.0

Query: Explain the applications of natural language processing.
P@5: 0.0
P@6: 0.0
P@10: 0.0

Query: How does robotics benefit from AI integration?
P@5: 0.0
P@6: 0.16666666666666666
P@10: 0.1

Query: What are the challenges in implementing AI in education?
P@5: 0.0
P@6: 0.0
P@10: 0.1

Query: How does data science contribute to climate change predictions?
P@5: 0.0
P@6: 0.0
P@10: 0.1

Query: What are the ethical considerations in AI development?
P@5: 0.0
P@6: 0.0
P@10: 0.0

Query: Explain the significance of AI in smart cities.
P@5: 0.2
P@6: 0.16666666666666666
P@10: 0.3

Query: How does AI enhance cybersecurity measures?
P@5: 0.0
P@6: 0.0
P@10: 0.2

Mean Average Precision (MAP): 0.023694255479969763
Mean Reciprocal Rank (MRR)

**Insights**

**Sparse Relevance:** Most queries have no relevant documents in the top K retrievals, indicating a challenge in accurately retrieving relevant information.

**Varied Precision:** Precision at different K values varies widely across queries, suggesting that the effectiveness of retrieval differs based on query characteristics.

**Mean Average Precision (MAP):** The overall MAP score is low (0.024), indicating that the system's average precision across all queries is modest.

**Mean Reciprocal Rank (MRR):** The MRR score (0.149) suggests that relevant documents are, on average, ranked relatively far down the list, indicating room for improvement in ranking accuracy.

**Room for Improvement:** The system may benefit from enhancements in relevance modeling, query expansion techniques, or algorithmic improvements to better capture document relevance.


**Overall**, the retrieval system's performance needs improvement, as indicated by low precision scores and modest MAP and MRR values. Strategies for enhancement include refining the document collection, optimizing algorithms, utilizing query expansion and relevance feedback, incorporating domain-specific knowledge, and adopting advanced evaluation metrics.


-------------------------------------------------------------------
- Part 2

In [12]:
# 0 is not relevant, 1 is relevant
annotator1_relevance = [1, 0, 1, 0, 1, 1, 0, 0, 1, 0]
annotator2_relevance = [1, 1, 1, 0, 1, 0, 0, 1, 1, 1]
annotator3_relevance = [1, 0, 0, 0, 1, 0, 0, 1, 1, 0]

In [13]:
## TODO: Implement your solution for part 2

from sklearn.metrics import cohen_kappa_score

# Compute pairwise Cohen's Kappa values
kappa_1_2 = cohen_kappa_score(annotator1_relevance, annotator2_relevance)
kappa_1_3 = cohen_kappa_score(annotator1_relevance, annotator3_relevance)
kappa_2_3 = cohen_kappa_score(annotator2_relevance, annotator3_relevance)

print("Cohen's Kappa between Annotator 1 and Annotator 2:", kappa_1_2)
print("Cohen's Kappa between Annotator 1 and Annotator 3:", kappa_1_3)
print("Cohen's Kappa between Annotator 2 and Annotator 3:", kappa_2_3)


Cohen's Kappa between Annotator 1 and Annotator 2: 0.19999999999999996
Cohen's Kappa between Annotator 1 and Annotator 3: 0.4
Cohen's Kappa between Annotator 2 and Annotator 3: 0.44444444444444453


**Insights**

**Cohen's Kappa Between Annotator 1 and Annotator 2:** The Kappa value of approximately 0.20 indicates slight agreement between annotator 1 and annotator 2. This suggests that their relevance assessments align only to a small extent beyond what would be expected by chance alone. There may be discrepancies or inconsistencies in their judgments, indicating the need for further examination or refinement of annotation guidelines.

**Cohen's Kappa Between Annotator 1 and Annotator 3:** With a Kappa value of 0.40, there is fair agreement between annotator 1 and annotator 3. This implies a moderate level of alignment in their relevance assessments, indicating that their judgments are more consistent compared to the previous pair. However, there is still room for improvement to achieve higher levels of agreement.

**Cohen's Kappa Between Annotator 2 and Annotator 3:**The Kappa value of approximately 0.44 indicates moderate agreement between annotator 2 and annotator 3. This suggests a relatively stronger alignment in their relevance assessments compared to the other pairs of annotators. Their judgments exhibit a more consistent pattern, although there may still be some areas of discrepancy.

**Overall**, while there are varying levels of agreement between pairs of annotators, none of the Kappa values indicate substantial or almost perfect agreement. This underscores the importance of assessing and addressing factors contributing to discrepancies in relevance assessments, such as clarifying annotation guidelines, providing additional training to annotators, or refining the annotation process to improve overall consistency and reliability.





